File: src/windows/recording.cpp
// SPDX-License-Identifier: LicenseRef-CASU-AntiCapitalist-1.4
// Full port of casu/recording.py MediaRecorder semantics (see header).
#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#define NOMINMAX
#define NOGDI
#include <windows.h>  // CREATE_NO_WINDOW: keep GUI child processes silent
#endif
#include "recording.hpp"

#include "casu/codec/ffprobe.hpp"
#include "casu/codec/tools.hpp"
#include "casu/formats.hpp"
#include "casu/json.hpp"
#include <QPointer>

#include <QDateTime>
#include <QDir>
#include <QFile>

#include <cmath>
#include <filesystem>
#include <thread>

namespace fs = std::filesystem;

namespace mpcasu {

namespace {

// casu/recording.py destination whitelist.
bool format_allowed(const QString& suffix) {
    static const QStringList allowed = {
        ".mkv", ".mp4", ".mov", ".ts", ".m2ts",
        ".webm", ".ogg", ".mp3", ".flac", ".wav"};
    return allowed.contains(suffix.toLower());
}

QString resolved_path(const QString& raw) {
    return QDir::cleanPath(QDir::current().absoluteFilePath(raw));
}

}  // namespace

RecordingController::RecordingController(QObject* parent) : QObject(parent) {}

RecordingController::~RecordingController() {
    if (proc_ && proc_->state() != QProcess::NotRunning) {
        proc_->terminate();
        if (!proc_->waitForFinished(3000)) proc_->kill();
    }
    cleanup_temp();
}

QString RecordingController::temporary_for(
    const QString& destination) const {
    const QFileInfo info(destination);
    const QString stem = info.completeBaseName();
    const QString suffix = info.suffix().isEmpty()
                               ? QString()
                               : QStringLiteral(".") + info.suffix();
    return info.absoluteDir().filePath(
        QStringLiteral(".%1.recording-%2%3")
            .arg(stem)
            .arg(QDateTime::currentMSecsSinceEpoch())
            .arg(suffix));
}

void RecordingController::cleanup_temp() {
    if (!temporary_.isEmpty() && QFile::exists(temporary_))
        QFile::remove(temporary_);
    temporary_.clear();
}

bool RecordingController::start(const QString& source,
                                const QString& destination, QString* error) {
    auto fail = [&](const char* msg) {
        if (error) *error = QString::fromLatin1(msg);
        state_ = State::Failed;
        if (on_state_changed) on_state_changed();
        return false;
    };
    // Source validation (reference ctor).
    if (source.isEmpty() || source.contains(u'\0') ||
        source.toUtf8().size() > 8192)
        return fail("recording source is invalid");
    // Destination format whitelist.
    const QFileInfo dest_info(destination);
    if (dest_info.suffix().isEmpty() ||
        !format_allowed(QStringLiteral(".") + dest_info.suffix()))
        return fail("recording destination format is unsupported");
    // Self-overwrite protection.
    if (resolved_path(source) == resolved_path(destination))
        return fail("recording cannot overwrite its source");

    QDir().mkpath(dest_info.absolutePath());
    if (proc_ && proc_->state() != QProcess::NotRunning) stop();

    const std::string exe = casu::codec::ffmpeg_path();
    if (exe.empty()) return fail("FFmpeg is required for recording");
    source_ = source;
    destination_ = QDir::cleanPath(destination);
    temporary_ = temporary_for(destination_);

    // Audio-only containers drop the video track (so recording a video to
    // MP3 works); video containers copy ALL streams (video + audio) so the
    // video picture is never switched off. Reference casu/recording.py.
    const QString suffix = dest_info.suffix().toLower();
    const bool audio_only = suffix == "mp3" || suffix == "ogg" ||
                            suffix == "flac" || suffix == "wav";
    QStringList args{QStringLiteral("-nostdin"),
                     QStringLiteral("-hide_banner"),
                     QStringLiteral("-loglevel"), QStringLiteral("error"),
                     QStringLiteral("-i"), source_};
    if (audio_only) {
        args << QStringLiteral("-map") << QStringLiteral("0:a:0?")
             << QStringLiteral("-vn");
        if (suffix == "mp3")
            args << QStringLiteral("-acodec") << QStringLiteral("libmp3lame")
                 << QStringLiteral("-q:a") << QStringLiteral("2");
        else if (suffix == "ogg")
            args << QStringLiteral("-acodec") << QStringLiteral("libvorbis")
                 << QStringLiteral("-q:a") << QStringLiteral("5");
        else if (suffix == "flac")
            args << QStringLiteral("-acodec") << QStringLiteral("flac");
        else if (suffix == "wav")
            args << QStringLiteral("-acodec") << QStringLiteral("pcm_s16le");
    } else {
        args << QStringLiteral("-map") << QStringLiteral("0")
             << QStringLiteral("-map_metadata") << QStringLiteral("0")
             << QStringLiteral("-map_chapters") << QStringLiteral("0")
             << QStringLiteral("-c") << QStringLiteral("copy");
    }
    args << QStringLiteral("-y") << temporary_;

    state_ = State::Starting;
    proc_ = new QProcess(this);
    proc_->setProgram(QString::fromStdString(exe));
    proc_->setArguments(args);
    connect(proc_, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
            this, &RecordingController::handle_finished);
#ifdef Q_OS_WIN
    // GUI app: never flash a console window for ffmpeg.
    proc_->setCreateProcessArgumentsModifier(
        [](QProcess::CreateProcessArguments* a) { a->flags |= CREATE_NO_WINDOW; });
#endif
    proc_->start();
    if (!proc_->waitForStarted(10000)) {
        cleanup_temp();
        return fail("could not start recording");
    }
    state_ = State::Recording;
    if (on_state_changed) on_state_changed();
    return true;
}

void RecordingController::stop() {
    if (!proc_ || proc_->state() == QProcess::NotRunning) return;
    state_ = State::Stopping;
    if (on_state_changed) on_state_changed();
    proc_->terminate();
    if (!proc_->waitForFinished(6000)) proc_->kill();
}

void RecordingController::kill() {
    if (proc_ && proc_->state() != QProcess::NotRunning) proc_->kill();
    cleanup_temp();
}

void RecordingController::handle_finished(int code,
                                          QProcess::ExitStatus status) {
    (void)status;
    // The muxed file and its probe are authoritative: a clean SIGTERM shows
    // up as exit 255 in some builds and must NOT mark the recording failed
    // (reference finish() semantics).
    publish_async();
}

void RecordingController::publish_async() {
    const QString temporary = temporary_;
    const QString destination = destination_;
    QPointer<RecordingController> guard(this);
    std::thread([guard, temporary, destination] {
        bool ok = false;
        QString detail;
        QString published;
        do {
            QFile tmp_file(temporary);
            if (!tmp_file.exists() || tmp_file.size() <= 0) {
                detail = QStringLiteral("recording produced no media");
                break;
            }
            casu::JsonValue probe;
            try {
                probe = casu::codec::probe_json(temporary.toStdString());
            } catch (const std::exception& exc) {
                detail = QStringLiteral("recording verification failed: %1")
                             .arg(QString::fromStdString(exc.what()));
                break;
            }
            bool playable = false;
            if (const casu::JsonValue* streams = probe.find("streams");
                streams && streams->is_array()) {
                for (const casu::JsonValue& s : streams->as_array().items) {
                    const casu::JsonValue* kind =
                        s.is_object() ? s.find("codec_type") : nullptr;
                    if (kind && kind->is_string() &&
                        (kind->as_string() == "audio" ||
                         kind->as_string() == "video")) {
                        playable = true;
                        break;
                    }
                }
            }
            if (!playable) {
                detail =
                    QStringLiteral("recording has no playable audio/video stream");
                break;
            }
            QFile::remove(destination);  // replace-existing rename
            std::error_code ec;
            fs::rename(temporary.toStdString(), destination.toStdString(), ec);
            if (ec) {
                detail = QStringLiteral("recording publish failed: %1")
                             .arg(QString::fromStdString(ec.message()));
                break;
            }
            ok = true;
            published = destination;
            detail = QStringLiteral("verified");
        } while (false);
        if (!ok) {
            QFile::remove(temporary);
        }
        if (!guard) return;
        QMetaObject::invokeMethod(guard, [guard, ok, published, detail] {
            if (!guard) return;
            guard->temporary_.clear();
            guard->state_ = State::Idle;
            if (guard->on_state_changed) guard->on_state_changed();
            if (guard->on_finished)
                guard->on_finished(published, ok, detail);
        });
    }).detach();
}

}  // namespace mpcasu

File: src/desktop/casu/recording.py
"""Atomic, fail-closed recording for MPCASU local and network sources."""
from __future__ import annotations

import os
import shutil
import subprocess
import tempfile
import time
from pathlib import Path

from .core import CasuError, ffprobe


class RecordingError(CasuError):
    pass


class MediaRecorder:
    """Record all available streams and publish only a verified final file."""

    def __init__(self, source: str | Path, destination: str | Path):
        self.source = str(source)
        if not self.source or "\0" in self.source or len(self.source.encode("utf-8")) > 8192:
            raise RecordingError("recording source is invalid")
        self.destination = Path(destination).expanduser().resolve()
        if not self.destination.suffix or self.destination.suffix.lower() not in {
                ".mkv", ".mp4", ".mov", ".ts", ".m2ts", ".webm", ".ogg", ".mp3", ".flac", ".wav"}:
            raise RecordingError("recording destination format is unsupported")
        try:
            if Path(self.source).expanduser().resolve() == self.destination:
                raise RecordingError("recording cannot overwrite its source")
        except (OSError, ValueError):
            pass
        self.destination.parent.mkdir(parents=True, exist_ok=True)
        fd, temporary = tempfile.mkstemp(
            prefix=f".{self.destination.stem}.recording-",
            suffix=self.destination.suffix, dir=self.destination.parent)
        os.close(fd); os.unlink(temporary)
        self.temporary = Path(temporary)
        self.process: subprocess.Popen | None = None
        self.started_at = 0.0

    @property
    def active(self) -> bool:
        return self.process is not None and self.process.poll() is None

    def start(self) -> None:
        if self.process is not None:
            raise RecordingError("recording has already been started")
        executable = shutil.which("ffmpeg")
        if not executable:
            raise RecordingError("FFmpeg is required for recording")
        command = self._build_command(executable)
        try:
            self.process = subprocess.Popen(command, stdin=subprocess.DEVNULL,
                                            stdout=subprocess.DEVNULL,
                                            stderr=subprocess.DEVNULL)
        except OSError as exc:
            self._cleanup()
            raise RecordingError(f"could not start recording: {exc}") from exc
        self.started_at = time.monotonic()

    def _build_command(self, executable: str) -> list[str]:
        """Build the ffmpeg command depending on target format.

        Video containers (mp4/mkv/webm/ts/mov/m2ts) stream-copy ALL streams so
        video AND audio are recorded and the video picture is never switched
        off. Audio-only containers (mp3/ogg/flac/wav) drop the video track and
        keep/transcode just the audio — recording a video to MP3 works.
        """
        suffix = self.destination.suffix.lower()
        audio_only = suffix in {".mp3", ".ogg", ".flac", ".wav"}
        command = [executable, "-nostdin", "-hide_banner", "-loglevel", "error",
                   "-i", self.source]
        if audio_only:
            # Keep audio only; transcode to the target codec inside the container.
            command += ["-map", "0:a:0?"]
            if suffix == ".mp3":
                command += ["-acodec", "libmp3lame", "-q:a", "2"]
            elif suffix == ".ogg":
                command += ["-acodec", "libvorbis", "-q:a", "5"]
            elif suffix == ".flac":
                command += ["-acodec", "flac"]
            elif suffix == ".wav":
                command += ["-acodec", "pcm_s16le"]
            command += ["-vn"]
        else:
            command += ["-map", "0", "-map_metadata", "0", "-map_chapters", "0",
                        "-c", "copy"]
        command += ["-y", str(self.temporary)]
        return command

    def finish(self, *, timeout: float = 10.0) -> Path:
        if self.process is None:
            raise RecordingError("recording was not started")
        process = self.process
        if process.poll() is None:
            process.terminate()
            try:
                process.wait(timeout=max(1.0, min(30.0, float(timeout))))
            except subprocess.TimeoutExpired:
                process.kill(); process.wait(timeout=5)
        # FFmpeg returns 255 on a clean SIGTERM in some builds. The actual
        # muxed file and probe are authoritative; a missing/invalid trailer is
        # rejected below.
        if not self.temporary.is_file() or self.temporary.stat().st_size <= 0:
            self._cleanup(); raise RecordingError("recording produced no media")
        try:
            probe = ffprobe(self.temporary)
            if not any(item.get("codec_type") in {"audio", "video"}
                       for item in probe.get("streams", [])):
                raise RecordingError("recording has no playable audio/video stream")
            os.replace(self.temporary, self.destination)
            try:
                directory = os.open(self.destination.parent, os.O_RDONLY)
                try: os.fsync(directory)
                finally: os.close(directory)
            except OSError:
                pass
        except (OSError, CasuError, ValueError) as exc:
            self._cleanup()
            if isinstance(exc, RecordingError): raise
            raise RecordingError(f"recording verification failed: {exc}") from exc
        return self.destination

    def abort(self) -> None:
        if self.process is not None and self.process.poll() is None:
            self.process.kill()
            try: self.process.wait(timeout=5)
            except subprocess.TimeoutExpired: pass
        self._cleanup()

    def _cleanup(self) -> None:
        try: self.temporary.unlink()
        except FileNotFoundError: pass

    def __del__(self):  # pragma: no cover - defensive process/file cleanup
        try:
            if self.active: self.abort()
        except Exception:
            pass
